File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {boolean}
 */
function inorder(root, inorderarr) {
    if (root == null) {
        return;
    }
    inorder(root.left, inorderarr);
    inorderarr.push(root.val);
    inorder(root.right, inorderarr);
}
var findTarget = function (root, k) {
    const inorderarr = []
    inorder(root, inorderarr);
    console.log(inorderarr)
    let low = 0;
    let high = inorderarr.length - 1;
    while (low < high) {
        if (inorderarr[low] + inorderarr[high] == k) {
            return true;
        }
        else if (inorderarr[low] + inorderarr[high] < k) {
            low++
        }
        else {
            high--;
        }
    }

    return false
};

File: 0653-two-sum-iv-input-is-a-bst/0653-two-sum-iv-input-is-a-bst.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    void morris_traversal(TreeNode* root, vector<int>& v) {
        // use simple morris travcersal here
        TreeNode *current = root;
        while(current!=NULL){
            if (current->left==NULL){
                v.push_back(current->val);
                current = current->right;
            }
            else {
                // current->left null nahi hai  toh left node kay right mai dekho agar vo right node kisi ko 
                // bi point nahi hai to usko current ko point karvao
                // agar vo current ko point hai toh iska matlab traverse ho chuka , print kar do, aur right mai jao
                TreeNode *leftNode = current->left;
                while(leftNode->right!=NULL && leftNode->right!= current){
                    leftNode = leftNode->right;
                }
                if (leftNode->right==current){
                    v.push_back(current->val);
                    leftNode->right = NULL;
                    current = current->right;
                }
                else{
                    leftNode->right= current;
                    current = current->left;
                }
            }
    }
    }
    bool findTarget(TreeNode* root, int k) {
        vector<int> v;
        morris_traversal(root, v);
        int i = 0;
        int j = v.size() - 1;
        while (j > i) {
            if (v[i] + v[j] == k) {
                return true;
            } else if (v[i] + v[j] < k) {
                i++;
            } else {
                j--;
            }
        }
        return false;
    }
};